File: Linked List/Singly Linked List/Singly Circular/SC.java
// Singly Circular Linked List

import java.util.Scanner;

class node
{
    public int data;
    public node next;

    public node(int no)
    {
        data =no;
        next =null;
    }
}

class Scircular
{
    public node head;
    public node tail;

    pblic Scircular()
    {
        head =null;
        tail=null;
    }

    public void InserFirst(int no)
    {
        node newn = new node(no);

        if(head == null && tail == null)
        {
            head = newn;
            tail=newn;
        }
        else
        {
            newn.next = head;
            head = newn;
        }
        tail.next = head;
    }

    public void Display()
    {
        node temp =head;

        while(temp != null)
        {
            System.out.println("|"+temp.data+"|->");
            temp = temp.next;
        }
    }

    public void InsertLast(int no)
    {
        node newn = new node(no);

        if(head == null && tail == null)
        {
            head = newn;
            tail =newn;
        }
        else
        {
            tail.next = newn;
            tail=newn;
        }
        tail.next =head;
    }

    public void DeleteFirst()
    {
        if(head == null && tail == null)
        {
            return 0;
        }
        else if( head == tail)
        {
            head = null;
            tail=null;
            System.gc();
        }
        else
        {
            head = head.next;
            System.gc();
            tail.next = head;
        }
    }

    public void DeleteLast()
    {
        if(head == null && tail == null)
        {
            return 0;
        }
        else if(head == tail)
        {
            head = null;
            tail=null;
            System.gc();
        }
        else
        {
            node temp=head;

            while(temp.next.next != tail)
            {
                temp =temp.next;
            }
            temp.next = tail;
            System.gc();
            tail.next =head;
        }
    }
}

class SC 
{
    public static void main(String a[])
    {
        Scanner sc = new Scaner(System.in);

        Scircular sc = new Scircular();

        System.out.println("Insert Element at First:");
        sc.InserFirst(30);
        sc.InserFirst(20);
        sc.InserFirst(10);
        sc.Display();
        System.out.println("\n----------------------------------------------");

        System.out.println("Insert Element at Last:");
        sc.InsertLast(300);
        sc.Display();
        System.out.println("\n----------------------------------------------");


        System.out.println("Delete First Element:");
        sc.DeleteFirst();
        sc.Display();
        System.out.println("\n----------------------------------------------");

        System.out.println("Delete Last Element:");
        sc.DeleteLast();
        sc.Display();
        System.out.println("\n----------------------------------------------");

    }
}

File: Linked List/Singly Linked List/Singly Circular/SC.c
#include<stdio.h>
#include<stdlib.h>

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node *PNODE;
typedef struct node **PPNODE;

int InsertFirst(PPNODE head,PPNODE tail,int no)
{
    PNODE newn=NULL;
    newn= (PNODE)malloc(sizeof(NODE));
    newn->data=no;
    newn->next =NULL;

    if((*head)== NULL && (*tail)==NULL)
    {
        *head =newn;
        *tail=newn;
    }
    else
    {
        newn->next =*head;
        *head = newn;
        (*tail)->next = *head;
    }
}

int InsertLast(PPNODE head,PPNODE tail,int no)
{
    PNODE newn=NULL;
    newn= (PNODE)malloc(sizeof(NODE));
    newn->data=no;
    newn->next =NULL;

    if((*head)== NULL && (*tail)==NULL)
    {
        *head =newn;
        *tail=newn;
    }
    else
    {
        (*tail)->next = newn;
        *tail=newn;
    }
    (*tail)->next = *head;
}

int DeleteFirst(PPNODE head,PPNODE tail)
{
    if((*head)== NULL && (*tail)==NULL)
    {
        return 0;
    }
    else if(*head == *tail)
    {
        free(*head);
        *head = NULL;
        *tail = NULL;
    }
    else
    {
        PNODE temp=*head;

        *head = (*head)->next;
        free(temp);
        (*tail)->next = *head;
    }   
}

int DeleteLast(PPNODE head,PPNODE tail)
{
    if((*head)== NULL && (*tail)==NULL)
    {
        return 0;
    }
    else if(*head == *tail)
    {
        free(*head);
        *head = NULL;
        *tail = NULL;
    }
    else
    {
        PNODE temp=*head;

        while(temp->next != *tail)
        {
            temp= temp->next;
        }
        free(temp->next);
        *tail= temp;
        (*tail)->next = *head;
    }   
}

int InsertAtPosition(PPNODE head,PPNODE tail,int no,int ipos)
{
    int iCountNode = Count(*head);

    if(ipos >= 1 || ipos < iCountNode+1)
    {
        return 0;
    }

    if(ipos ==1)
    {
        InsertFirst(head,tail,no);
    }
    else if(ipos == iCountNode+1)
    {
       InsertLast(head,tail,no);
    }
    else
    {
        PNODE temp=*head;

        for(int i=0; i<ipos-1; i++)
        {
            temp= temp->next;
        }
        newn->next = temp->next;
        temp->next = newn;
    }   
}

int DeleteAtPosition(PPNODE head,PPNODE tail,int ipos)
{
    int iCountNode = Count(*head);

    if(ipos >= 1 || ipos < iCountNode)
    {
        return 0;
    }

    if(ipos ==1)
    {
        DeleteFirst(head,tail);
    }
    else if(ipos == iCountNode)
    {
       DeleteLast(head,tail);
    }
    else
    {
        PNODE temp=*head;
        PNODE target=NULL;

        for(int i=0; i<ipos-1; i++)
        {
            temp= temp->next;
        }
        target = temp->next;
        temp->next = target->next;
        free(target);
    }   
}

int Display(PNODE head,PNODE tail)
{
    PNODE temp=head;
    do 
    {
        printf("|%d|->",temp->data);
        temp=temp->next;
    }while(temp != tail->next);
}

int main()
{
    PNODE first=NULL;
    PNODE last =NULL;
    int ipos=0,idel=0;

    InsertFirst(&first,&last,30);
    InsertFirst(&first,&last,20);
    InsertFirst(&first,&last,10);
    printf("Insert Elements at First:");
    Display(first,last);

    printf("\n\n--------------------------------------------------");
    InsertLast(&first,&last,80);
    printf("\nInsert Elements at Last:");
    Display(first,last);

    printf("\n\n--------------------------------------------------");
    DeleteFirst(&first,&last);
    printf("\nDelete First Element:");
    Display(first,last);

    printf("\n\n--------------------------------------------------");
    DeleteLast(&first,&last);
    printf("\nDelete Last Element:");
    Display(first,last);

    printf("\n\n--------------------------------------------------");
    InsertAtPosition(&first,&last,ipos,90);
    printf("\nEnter position:");
    scanf("%d",&ipos);
    printf("\nInsett Element at Position:");
    Display(first,last);

    printf("\n\n--------------------------------------------------");
    DeleteAtPosition(&first,&last,idel);
    printf("\nEnter position:");
    scanf("%d",&idel);
    printf("\nDelete Element at Position:");
    Display(first,last);

    return 0;
}